package/cell_modem_daemon/src/at_command_utils.c

```diff
@@ -0,0 +1,336 @@
+         > 0) {
+    current += num_read;
+    if (current[-1] == '\n' || current[-1] == '\r') {
+      /* null terminate the string and see if we got an OK response */
+      *current = '\0';
+      if (strstr(current - 4, "OK") != NULL
+          || strstr(current - 7, "ERROR") != NULL) {
+        at_command->complete = true;
+      }
+    }
+    if (at_command->complete || current == eob - 1) {
+      break;
+    }
+  }
+
+  *current = '\0'; // Null terminate in case we missed all returns
+  at_command->result_length = (size_t)(current - at_command->result);
+}
+
+void at_serial_port_execute_command(at_serial_port_t *port,
+                                    at_serial_port_command_t *at_command)
+{
+  at_serial_port_open(port);
+  if (at_serial_port_is_open(port)) {
+    snprintf(at_command->result,
+             sizeof(at_command->result),
+             "%s returned ",
+             at_command->command);
+
+    int n = (int)write(port->fd, at_command->command, strlen(at_command->command));
+    n += (int)write(port->fd, "\r", 1);
+
+    if (n == (int)strlen(at_command->command) + 1) {
+      at_serial_port_wait_command_response(port, at_command);
+      if (!at_command->complete) {
+        piksi_log(
+          LOG_ERR,
+          "Unable to resolve result of command '%s' to port '%s'. %d characters returned",
+          at_command->command,
+          port->port_name,
+          at_command->result_length);
+      }
+    } else {
+      piksi_log(LOG_ERR,
+                "Unable to write command '%s' to port '%s'",
+                at_command->command,
+                port->port_name);
+    }
+
+    at_serial_port_close(port);
+  } else {
+    piksi_log(LOG_ERR, "Unable to open port: %s", port->port_name);
+  }
+}
+
+const char *at_serial_port_command_result(at_serial_port_command_t *at_command)
+{
+  return at_command->result;
+}
+
+/* The following from Telit HE910 AT Command Reference
+ * +CSQ: <rssi>,<ber>
+ * where
+ * <rssi> - received signal strength indication
+ *  0 - (-113) dBm or less
+ *  1 - (-111) dBm
+ *  2..30 - (-109)dBm..(-53)dBm / 2 dBm per step
+ *  31 - (-51)dBm or greater
+ *  99 - not known or not detectable
+ * <ber> - bit error rate (in percent)
+ *  0 - less than 0.2%
+ *  1 - 0.2% to 0.4%
+ *  2 - 0.4% to 0.8%
+ *  3 - 0.8% to 1.6%
+ *  4 - 1.6% to 3.2%
+ *  5 - 3.2% to 6.4%
+ *  6 - 6.4% to 12.8%
+ *  7 - more than 12.8%
+ *  99 - not known or not detectable
+ */
+static const s8 signal_to_dbm_lookup[] = {
+  -113, // zero case
+  -111, -109, -107, -105, -103, -101, -99, -97, -95, -93,  // 1 - 30 cases
+  -91,  -89,  -87,  -85,  -83,  -81,  -79,  -77, -75, -73,
+  -71, -69,  -67,  -65,  -63,  -61,  -59,  -57,  -55, -53,
+  -51 // 31 case
+};
+
+static s8 modem_signal_strength_to_dbm(u32 signal_raw)
+{
+  s8 signal_dbm = 0; // unknown will remain 0
+
+  if (signal_raw < sizeof(signal_to_dbm_lookup) / sizeof(signal_to_dbm_lookup[0])) {
```

Comment:
I tried to introduce a macro for this in #461 (which wasn't merged), I think now's a good time to introduce it in ` package/libpiksi/libpiksi/include/libpiksi/util.h`:

```c
#define COUNT_OF(x) ((sizeof(x)/sizeof(0[x])) / ((size_t)(!(sizeof(x) % sizeof(0[x])))))
```

(see https://stackoverflow.com/a/4415646 for an explanation of the `0[x]` stuff)